peer.go

```diff
@@ -4,134 +4,291 @@ import (
 		func() {
 			var n int
-			wire.WriteBinary(ourNodeInfo, conn, &n, &err1)
+			wire.WriteBinary(ourNodeInfo, p.conn, &n, &err1)
 		},
 		func() {
 			var n int
-			wire.ReadBinary(peerNodeInfo, conn, maxNodeInfoSize, &n, &err2)
+			wire.ReadBinary(peerNodeInfo, p.conn, maxNodeInfoSize, &n, &err2)
 			log.Notice("Peer handshake", "peerNodeInfo", peerNodeInfo)
 		})
 	if err1 != nil {
-		return nil, err1
+		return err1
 	}
 	if err2 != nil {
-		return nil, err2
+		return err2
 	}
-	peerNodeInfo.RemoteAddr = conn.RemoteAddr().String()
-	return peerNodeInfo, nil
-}
 
-// NOTE: call peerHandshake on conn before calling newPeer().
-func newPeer(config cfg.Config, conn net.Conn, peerNodeInfo *NodeInfo, outbound bool, reactorsByCh map[byte]Reactor, chDescs []*ChannelDescriptor, onPeerError func(*Peer, interface{})) *Peer {
-	var p *Peer
-	onReceive := func(chID byte, msgBytes []byte) {
-		reactor := reactorsByCh[chID]
-		if reactor == nil {
-			PanicSanity(Fmt("Unknown channel %X", chID))
+	if p.authEnc {
+		// Check that the professed PubKey matches the sconn's.
+		if !peerNodeInfo.PubKey.Equals(p.PubKey()) {
+			return fmt.Errorf("Ignoring connection with unmatching pubkey: %v vs %v",
+				peerNodeInfo.PubKey, p.PubKey())
 		}
-		reactor.Receive(chID, p, msgBytes)
 	}
-	onError := func(r interface{}) {
-		p.Stop()
-		onPeerError(p, r)
+
+	// Remove deadline
+	p.conn.SetDeadline(time.Time{})
+
+	peerNodeInfo.RemoteAddr = p.RemoteAddr().String()
+
+	p.NodeInfo = peerNodeInfo
+	p.Key = peerNodeInfo.PubKey.KeyString()
+
+	return nil
+}
+
+// RemoteAddr returns the remote network address.
+func (p *Peer) RemoteAddr() net.Addr {
+	return p.conn.RemoteAddr()
+}
+
+// PubKey returns the remote public key.
+func (p *Peer) PubKey() crypto.PubKeyEd25519 {
+	if p.authEnc {
+		return p.conn.(*SecretConnection).RemotePubKey()
 	}
-	mconn := NewMConnection(config, conn, chDescs, onReceive, onError)
-	p = &Peer{
-		outbound: outbound,
-		mconn:    mconn,
-		NodeInfo: peerNodeInfo,
-		Key:      peerNodeInfo.PubKey.KeyString(),
-		Data:     NewCMap(),
+	if p.NodeInfo == nil {
+		panic("Attempt to get peer's PubKey before calling Handshake")
 	}
-	p.BaseService = *NewBaseService(log, "Peer", p)
-	return p
+	return p.PubKey()
 }
 
+// OnStart implements BaseService.
 func (p *Peer) OnStart() error {
 	p.BaseService.OnStart()
 	_, err := p.mconn.Start()
 	return err
 }
 
+// OnStop implements BaseService.
 func (p *Peer) OnStop() {
 	p.BaseService.OnStop()
 	p.mconn.Stop()
 }
 
+// Connection returns underlying MConnection.
 func (p *Peer) Connection() *MConnection {
 	return p.mconn
 }
 
+// IsOutbound returns true if the connection is outbound, false otherwise.
 func (p *Peer) IsOutbound() bool {
 	return p.outbound
 }
 
+// Send msg to the channel identified by chID byte. Returns false if the send
+// queue is full after timeout, specified by MConnection.
 func (p *Peer) Send(chID byte, msg interface{}) bool {
 	if !p.IsRunning() {
+		// FIXME shouldn't we panic here?
```

Comment:
Don't think so. We do seem to remove the peer before we call Stop, but see sw.Broadcast, where we fetch the list of peers and loop over them - while we're looping, one peer may be removed and stopped, in which case the Send will fail.